### PR TITLE
Order

### PR DESCRIPTION
**Pull request by isidzukuri**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -4,6 +4,10 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   before_action :messages_for_user, only: [:index, :show, :new, :edit]
 
+  def after_sign_in_path_for(_resource)
```
Comment (AKovtunov) [inline app/controllers/application_controller.rb]: I think you can find in devise smth for this

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      @order.order_items << OrderItem.new(product_id: product, quantity: 1)
+    end
+  end
+
+  def force_login
+    session[:previous_url] = basket_index_path
+    redirect_to new_user_session_path
+  end
+
+  def successful_creation
```
Comment (AKovtunov) [inline app/controllers/orders_controller.rb]: probably it should be repo

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    add_flash 'Order created successfully'
+    redirect_to @order
+  end
+
+  def failed_creation
+    add_flash 'Order not valid', :error
+    redirect_to basket_index_path
+  end
+
+  def products
```
Comment (AKovtunov) [inline app/controllers/orders_controller.rb]: don't use session here

File: app/models/order.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+class Order < ApplicationRecord
+  resourcify
+  belongs_to :order_status
```
Comment (AKovtunov) [inline app/models/order.rb]: `belongs_to :order_status` why? why not just a column?

File: app/models/order.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  validate :user_present
+
+  def subtotal
+    order_items.collect(&:total_price).sum
+  end
+
+  private
+
+  def set_order_status
+    self.order_status_id = 1
```
Comment (AKovtunov) [inline app/models/order.rb]: take a look at `enum`

File: app/models/order_item.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    errors.add(:product, 'is not valid.') if product.nil?
+  end
+
+  def order_present
+    errors.add(:order, 'is not a valid order.') if order.nil?
+  end
+
+  def finalize
+    self[:unit_price] = unit_price
+    self[:total_price] = quantity * self[:unit_price]
```
Comment (AKovtunov) [inline app/models/order_item.rb]: you are duplicating it in `total price`

**Comment by isidzukuri** [comment]
@AKovtunov  Its not completed version.  Ill let u know when it be.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (isidzukuri) [inline app/controllers/application_controller.rb]: its a solution from a device wiki. There are a lot of examples, but they looks sophisticated. https://github.com/plataformatec/devise/wiki/How-To:-redirect-to-a-specific-page-on-successful-sign-in

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (isidzukuri) [inline app/controllers/orders_controller.rb]: not excluded

File: app/models/order.rb (hunk 4 above)
Comment (isidzukuri) [inline app/models/order.rb]: looks more flexible for me. Statuses have names whitch should be stored somewhere. They can be stored in Order model like a hash in constant or in some translation file for example. But i dont believe thats is a right way. Also if statuses will have their own model there will be ability to add some other attributes. Example: description or enabled/disabled for usage.

File: app/models/order.rb (hunk 5 above)
Comment (isidzukuri) [inline app/models/order.rb]: fixed

File: app/models/order_item.rb (hunk 6 above)
Comment (isidzukuri) [inline app/models/order_item.rb]: removed

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (isidzukuri) [inline app/controllers/orders_controller.rb]: basket functionality moved to repo

**Pull request merged**